pkg/template/resource.go

```diff
@@ -46,40 +47,58 @@ type getClusterTriggerBinding func(name string, options metav1.GetOptions) (*tri
-	tb := make([]*triggersv1.TriggerBinding, 0, len(trigger.Bindings))
-	ctb := make([]*triggersv1.ClusterTriggerBinding, 0, len(trigger.Bindings))
-	for _, b := range trigger.Bindings {
-		if b.Spec != nil {
-			tb = append(tb, &triggersv1.TriggerBinding{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: b.Name,
-				},
-				Spec: *b.Spec,
+	bp, err := resolveBindingsToParams(trigger.Bindings, getTB, getCTB)
+	if err != nil {
+		return ResolvedTrigger{}, fmt.Errorf("failed to resolve bindings: %w", err)
+	}
+	tt, err := getTT(trigger.Template.Name, metav1.GetOptions{})
+	if err != nil {
+		return ResolvedTrigger{}, fmt.Errorf("error getting TriggerTemplate %s: %w", trigger.Template.Name, err)
+	}
+	return ResolvedTrigger{TriggerTemplate: tt, BindingParams: bp}, nil
+}
+
+// resolveBindingsToParams takes in both embedded bindings and references and returns a list of resolved Param values.ResolveBindingsToParams
+func resolveBindingsToParams(bindings []*triggersv1.TriggerSpecBinding, getTB getTriggerBinding, getCTB getClusterTriggerBinding) ([]triggersv1.Param, error) {
+	bindingParams := []triggersv1.Param{}
+	for _, b := range bindings {
+		switch {
+		case b.Spec != nil: // Could also call SetDefaults and not rely on this?
+			bindingParams = append(bindingParams, b.Spec.Params...)
+
+		case b.Name != "" && b.Value != nil:
+			bindingParams = append(bindingParams, triggersv1.Param{
+				Name:  b.Name,
+				Value: *b.Value,
 			})
-			continue
-		}
 
-		if b.Kind == triggersv1.ClusterTriggerBindingKind {
-			ctb2, err := getCTB(b.Ref, metav1.GetOptions{})
+		case b.Ref != "" && b.Kind == triggersv1.ClusterTriggerBindingKind:
+			ctb, err := getCTB(b.Ref, metav1.GetOptions{})
 			if err != nil {
-				return ResolvedTrigger{}, fmt.Errorf("error getting ClusterTriggerBinding %s: %w", b.Name, err)
+				return nil, fmt.Errorf("error getting ClusterTriggerBinding %s: %w", b.Name, err)
 			}
-			ctb = append(ctb, ctb2)
-		} else {
-			tb2, err := getTB(b.Ref, metav1.GetOptions{})
+			bindingParams = append(bindingParams, ctb.Spec.Params...)
+
+		case b.Ref != "": // if no kind is set, assume NamespacedTriggerBinding
+			tb, err := getTB(b.Ref, metav1.GetOptions{})
 			if err != nil {
-				return ResolvedTrigger{}, fmt.Errorf("error getting TriggerBinding %s: %w", b.Name, err)
+				return nil, fmt.Errorf("error getting TriggerBinding %s: %w", b.Name, err)
 			}
-			tb = append(tb, tb2)
+			bindingParams = append(bindingParams, tb.Spec.Params...)
+		default:
+			return nil, fmt.Errorf("invalid binding: %v", b)
 		}
 	}
 
-	ttName := trigger.Template.Name
-	tt, err := getTT(ttName, metav1.GetOptions{})
-	if err != nil {
-		return ResolvedTrigger{}, fmt.Errorf("error getting TriggerTemplate %s: %w", ttName, err)
+	// Check for duplicate params
+	seen := make(map[string]bool, len(bindingParams))
+	for _, p := range bindingParams {
+		if seen[p.Name] {
+			return nil, fmt.Errorf("duplicate param name: %s", p.Name)
```

Comment:
This looks to be a new behaviour. Should we document this as breaking change?

Reply:
we already had this check...just in a different function before: https://github.com/tektoncd/triggers/blob/v0.8.1/pkg/template/resource.go#L151